### DISCUSSION TITLE
Module not found: Can't resolve 'zustand/middleware/devtools' in Zustand v5

### DISCUSSION THREAD

**Discussion by frixaco** [pmndrs/zustand#2797] [Bug report] [answered]
### Bug Description

Just upgraded Zustand to v5 in our Next.js project and got this error for both `bun dev` and `bun run build`:

```
❯ bun run build
$ next build
  ▲ Next.js 14.2.15

   Creating an optimized production build ...
Failed to compile.

./app/useStore.tsx
Module not found: Can't resolve 'zustand/middleware/devtools'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./app/page.tsx
```

Setup:
```tsx
import { devtools } from "zustand/middleware/devtools";
import { createWithEqualityFn } from "zustand/traditional";
import { shallow } from "zustand/vanilla/shallow";

type Store = {
  userId: string | null;
};

export const useUserStore = createWithEqualityFn<Store>()(
  devtools<Store>(
    () => ({
      userId: null,
    }),
    {
      name: "my-devtools",
    },
  ),
  shallow,
);
```

### Reproduction Link

https://github.com/frixaco/repro-zustand-nextjs-devtools

**Comment by dai-shi** [comment]
Thanks for reporting.

Can anyone identify if this is bun-only issue?

**Comment by frixaco** [comment]
Thank for a quick response @dai-shi!

Doesn't seem so.
Just tried creating new Next.js app using NPM and copied over [page.tsx](https://github.com/frixaco/repro-zustand-nextjs-devtools/blob/main/app/page.tsx) and [useStore.tsx](https://github.com/frixaco/repro-zustand-nextjs-devtools/blob/main/app/useStore.tsx). Same error.

NPM 10.8.2
Node 20.18.0

**Reply by dai-shi** [reply]
thanks. can you make sure that this doesn't happen with the latest version of v4?

**Reply by frixaco** [reply]
This doesn't happen with latest v4 version - `4.5.5`. 
(created fresh Next.js app using NPM, same code but had to import `devtools` from `zustand/middleware` instead of `zustand/middleware/devtools`)

**Reply by frixaco** [reply]
sorry, I just realized if I import from `zustand/middleware` in v5, it works as well. Is this the supposed way to import in v5?

**Reply by dai-shi** [reply]
Ohhh! I see. I didn't notice it. Yeah, `import { devtools } from 'zustand/middleware'` should work for both v4 and v5. I wasn't aware v4 can import it differently. Maybe something was loose in v4.

**Reply by frixaco** [reply]
@dai-shi 

Just did a bit testing (including vanilla JS). Basically:
- `zustand/middleware/<some name>` - `devtools` throws an error (v4 throws a TS error right inside VSCode, v5 only during `bun run dev/build`), all other middleware work - both in v4 and v5
- `zustand/middleware` - all middleware work - both in v4 and v5

Docs currently use `zustand/middleware/<some name>` for all [middleware](https://zustand.docs.pmnd.rs/middlewares/devtools#debugging-a-slices-pattern-based-store) except `persist`.
So I guess it's either a bug with devtools or docs just need to be updated to use `zustand/middleware` at least for devtools?

**Reply by dai-shi** [reply]
Ah, that's nice catch. This is my oversight and it's the doc issue.
We should import all middleware from `zustand/middleware` except for `zustand/middleware/immer` which depends on an external library.
@dbritto-dev Can you fix docs please?

**Reply by dbritto-dev** [reply]
@dai-shi sure

**Reply by dbritto-dev** [reply, marked as answer]
@frixaco @dai-shi here you go -> https://github.com/pmndrs/zustand/pull/2798
